components/script/dom/idbcursor.rs

```diff
@@ -0,0 +1,163 @@
+        got_value: bool,
+        source: ObjectStoreOrIndex,
+        range: IndexedDBKeyRange,
+        key_only: bool,
+    ) -> IDBCursor {
+        IDBCursor {
+            reflector_: Reflector::new(),
+            transaction,
+            range,
+            source,
+            direction,
+            position: DomRefCell::new(None),
+            key: DomRefCell::new(None),
+            value: Heap::default(),
+            got_value: Cell::new(got_value),
+            object_store_position: DomRefCell::new(None),
+            key_only,
+            request: Default::default(),
+        }
+    }
+
+    #[expect(dead_code)]
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        global: &GlobalScope,
+        transaction: DomRoot<IDBTransaction>,
+        direction: IDBCursorDirection,
+        got_value: bool,
+        source: ObjectStoreOrIndex,
+        range: IndexedDBKeyRange,
+        key_only: bool,
+        can_gc: CanGc,
+    ) -> DomRoot<IDBCursor> {
+        reflect_dom_object(
+            Box::new(IDBCursor::new_inherited(
+                transaction,
+                direction,
+                got_value,
+                source,
+                range,
+                key_only,
+            )),
+            global,
+            can_gc,
+        )
+    }
+
+    pub(crate) fn get_value(&self) -> JSVal {
+        self.value.get()
+    }
+
+    /// <https://www.w3.org/TR/IndexedDB-2/#cursor-effective-key>
+    pub(crate) fn effective_key(&self) -> Option<IndexedDBKeyType> {
+        match &self.source {
+            ObjectStoreOrIndex::ObjectStore(_) => self.position.borrow().clone(),
+            ObjectStoreOrIndex::Index(_) => self.object_store_position.borrow().clone(),
+        }
+    }
+}
+
+impl IDBCursorMethods<crate::DomTypeHolder> for IDBCursor {
+    /// <https://www.w3.org/TR/IndexedDB-2/#dom-idbcursor-source>
+    fn Source(&self) -> IDBObjectStoreOrIDBIndex {
+        match &self.source {
+            ObjectStoreOrIndex::ObjectStore(source) => {
+                IDBObjectStoreOrIDBIndex::IDBObjectStore(source.clone())
+            },
+            ObjectStoreOrIndex::Index(source) => IDBObjectStoreOrIDBIndex::IDBIndex(source.clone()),
+        }
+    }
+
+    /// <https://www.w3.org/TR/IndexedDB-2/#dom-idbcursor-direction>
+    fn Direction(&self) -> IDBCursorDirection {
+        self.direction
+    }
+
+    /// <https://www.w3.org/TR/IndexedDB-2/#dom-idbcursor-key>
+    fn Key(&self, cx: SafeJSContext, value: MutableHandleValue) {
+        key_type_to_jsval(cx, self.key.borrow().as_ref().unwrap(), value);
+    }
+
+    /// <https://www.w3.org/TR/IndexedDB-2/#dom-idbcursor-primarykey>
+    fn PrimaryKey(&self, cx: SafeJSContext, value: MutableHandleValue) {
+        key_type_to_jsval(cx, &self.effective_key().unwrap(), value);
+    }
+
+    /// <https://w3c.github.io/IndexedDB/#dom-idbcursor-request>
+    fn Request(&self) -> DomRoot<IDBRequest> {
+        self.request.get().unwrap()
```

Comment:
I would use `expect("IDBCursor.request not set")` instead of unwrap with proper panic message